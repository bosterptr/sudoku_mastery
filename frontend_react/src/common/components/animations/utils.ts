import { CSSProperties } from 'react';

export const reflow = (node: Element) => node.scrollTop;

interface ComponentProps {
  easing: string | { enter?: string; exit?: string } | undefined;
  style: CSSProperties | undefined;
  timeout: number | { enter?: number; exit?: number };
}

interface Options {
  mode: 'enter' | 'exit';
}

interface TransitionProps {
  duration: string | number;
  easing: string | undefined;
  delay: string | undefined;
}

export function getTransitionProps(props: ComponentProps, options: Options): TransitionProps {
  const { timeout, easing, style = {} } = props;

  return {
    duration:
      style.transitionDuration ??
      (typeof timeout === 'number' ? timeout : timeout[options.mode] || 0),
    easing:
      style.transitionTimingFunction ??
      (typeof easing === 'object' ? easing[options.mode] : easing),
    delay: style.transitionDelay,
  };
}

export function getAutoHeightDuration(height: number) {
  if (!height) return 0;
  const constant = height / 36;
  // https://www.wolframalpha.com/input/?i=(4+%2B+15+*+(x+%2F+36+)+**+0.25+%2B+(x+%2F+36)+%2F+5)+*+10
  return Math.round((4 + 15 * constant ** 0.25 + constant / 5) * 10);
}
