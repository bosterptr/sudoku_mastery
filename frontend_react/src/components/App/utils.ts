const GL_BUFFER_STATUS_CHECK_INTERVAL = 2;

export type IOpenglSetup = [
  OffscreenCanvas,
  WebGLProgram,
  WebGLProgram,
  WebGLProgram,
  ITexFbPair,
  ITexFbPair,
  number,
];

export interface ITexFbPair {
  tex: WebGLTexture;
  fb: WebGLFramebuffer;
}
export const getGrayScaleFromUint8Array = (
  dataArray: Uint8Array,
  width: number,
  height: number,
  cachedGrayScaleUintArray: Uint8Array,
  cachedGrayScaleSize: number,
): Uint8Array => {
  if (cachedGrayScaleSize !== width * height) {
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray = new Uint8Array(width * height);
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleSize = width * height;
  }
  const { length } = dataArray;
  for (let i = 0, luminanceIndex = 0; i < length; i += 4, luminanceIndex += 1) {
    // eslint-disable-next-line no-bitwise
    const r = dataArray[i] >> 2;
    // eslint-disable-next-line no-bitwise
    const g = dataArray[i + 1] >> 1;
    // eslint-disable-next-line no-bitwise
    const b = dataArray[i + 2] >> 3;
    // const luminance = 0.299 * r + 0.587 * g + 0.114 * b;
    // luminanceData[luminanceIndex] =
    //   0.299 * dataArray[i] +
    //   0.587 * dataArray[i + 1] +
    //   0.114 * dataArray[i + 2];
    // eslint-disable-next-line no-param-reassign
    cachedGrayScaleUintArray[luminanceIndex] = r + b + g;
  }
  return cachedGrayScaleUintArray;
};

export const clientWaitAsync = (
  gl: WebGL2RenderingContext,
  sync: WebGLSync,
  flags = 0,
  intervalMs = 10,
) => {
  return new Promise<void>((resolve, reject) => {
    const check = () => {
      const res = gl.clientWaitSync(sync, flags, 0);
      if (res === gl.WAIT_FAILED) {
        reject();
        return;
      }
      if (res === gl.TIMEOUT_EXPIRED) {
        setTimeout(check, intervalMs);
        return;
      }
      resolve();
    };
    check();
  });
};

export const readPixelsAsync = (
  gl: WebGL2RenderingContext,
  width: number,
  height: number,
  buffer: Uint8Array,
): Promise<Uint8Array> => {
  const bufpak = gl.createBuffer();
  gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
  gl.bufferData(gl.PIXEL_PACK_BUFFER, buffer.byteLength, gl.STREAM_READ);
  gl.readPixels(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, 0);
  const sync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
  if (!sync) throw new Error('WebGLSync is empty');
  gl.flush();
  return clientWaitAsync(gl, sync, 0, GL_BUFFER_STATUS_CHECK_INTERVAL).then(() => {
    gl.deleteSync(sync);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, bufpak);
    gl.getBufferSubData(gl.PIXEL_PACK_BUFFER, 0, buffer);
    gl.bindBuffer(gl.PIXEL_PACK_BUFFER, null);
    gl.deleteBuffer(bufpak);
    return buffer;
  });
};

export function setAttributes(gl: WebGL2RenderingContext, program: WebGLProgram, name: string) {
  const location = gl.getAttribLocation(program, name);
  gl.enableVertexAttribArray(location);
  gl.vertexAttribPointer(location, 2, gl.FLOAT, false, 0, 0);
}

export function isMobile() {
  const userAgent = navigator.userAgent.toLowerCase();
  return /iphone|ipod|android|ie|blackberry|fennec/.test(userAgent);
}
export const getCameraResolution = () => {
  if (isMobile()) {
    return {
      width: { ideal: document.body.clientWidth > 500 ? 500 : document.body.clientWidth },
      height: { ideal: document.body.clientHeight > 500 ? 500 : document.body.clientHeight },
    };
  }
  return {
    width: { ideal: 1000 },
    height: { ideal: 1000 },
  };
};
