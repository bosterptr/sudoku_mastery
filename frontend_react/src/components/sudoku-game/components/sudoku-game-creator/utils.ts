import { ALL_SUDOKU_CELLS, CANVAS_SIZE, SUDOKU_CELLS_IN_ROW } from 'app/core/constants/sudoku';
import { ICellNotes, IMission } from './types';

export const generateNotes = (): ICellNotes[] => {
  const newArray: ICellNotes[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    newArray.push([]);
  }
  return newArray;
};
export const getBlockSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 2 * pixelRatio) / 3;
export const getAdjustedPixelRatio = () => {
  const { devicePixelRatio } = window;
  return devicePixelRatio > 2 ? 2 * devicePixelRatio : 4;
};
export const getCellSize = (pixelRatio: number) => (CANVAS_SIZE * pixelRatio - 4 * pixelRatio) / 9;
export const getContextSize = (pixelRatio: number) => CANVAS_SIZE * pixelRatio;
export const getOuterSize = (pixelRatio: number) => 2 * pixelRatio;
export const getThinSize = (pixelRatio: number) => 0.75 * pixelRatio;
export const getX = (index: number) => Math.floor(index / 9);
export const getY = (index: number) => index % 9;
export const getCellXPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getY(index);
export const getCellYPosition = (outerSize: number, cellSize: number, index: number) =>
  outerSize + cellSize * getX(index);
export const getCubeId = (index: number) =>
  `${Math.floor(getX(index) / 3)}${Math.floor(getY(index) / 3)}`;
export const EMPTY_CELL = '.';
export const isSudokuCellEmpty = (char: string) => char === EMPTY_CELL;
export const generateArrWithTrueAtIndexes = (indexes: number[]): boolean[] => {
  const newArray: boolean[] = [];
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (indexes.includes(cellIndex)) newArray[cellIndex] = true;
    else newArray[cellIndex] = false;
  }
  return newArray;
};
export const getScoreFromMove = (values: IMission, solution: string, selectedCell: number) => {
  let fullRows = 0;
  let fullColumns = 0;
  let fullCubes = 0;
  for (let cellIndex = 0; cellIndex < ALL_SUDOKU_CELLS; cellIndex += 1) {
    if (values[cellIndex] === solution[cellIndex]) {
      if (getY(cellIndex) === getY(selectedCell)) fullRows += 1;
      if (getX(cellIndex) === getX(selectedCell)) fullColumns += 1;
      if (getCubeId(cellIndex) === getCubeId(selectedCell)) fullCubes += 1;
    }
  }
  let score = 1;
  if (fullRows === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullColumns === SUDOKU_CELLS_IN_ROW) score += 5;
  if (fullCubes === 9) score += 5;
  return score;
};
