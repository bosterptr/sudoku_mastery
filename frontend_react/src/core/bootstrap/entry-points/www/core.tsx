/* eslint-disable prettier/prettier */
// import '../../../../wdyr';
import { bootstrap } from 'app/core/bootstrap/common';
import { DefaultRoot } from 'app/pages/root';

// eslint-disable-next-line camelcase
bootstrap(DefaultRoot);
