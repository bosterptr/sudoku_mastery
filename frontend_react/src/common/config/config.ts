import { Locale, LocaleData } from 'app/core/i18n';
import { IBuildType } from '../models/build_type';

export abstract class Config {
  public abstract apiBaseURL: string;

  public abstract baseURL: string;

  public abstract buildType: IBuildType;

  public optimizedBuild = process.env.NODE_ENV === 'production';

  public authSettings = {
    /** Name of the cookie to store user authentication in. */
    cookieName: 'auth',
  };

  public locales: Locale[] = [
    {
      name: 'English',
      languageCode: 'en',
      locale: 'en-US',
      default: true,
    },
    {
      name: 'Polski',
      languageCode: 'pl',
      locale: 'pl-PL',
      loader: () =>
        import(
          /* webpackChunkName: 'core.locales.pl-PL' */ 'app/core/i18n/locales/pl-PL'
        ) as unknown as Promise<LocaleData>,
    },
  ];
}
