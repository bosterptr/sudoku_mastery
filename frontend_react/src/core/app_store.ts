/* eslint-disable no-underscore-dangle */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/ban-types */
import { configureStore, Slice, SliceCaseReducers } from '@reduxjs/toolkit';
import { Config } from 'app/common/config';
import { AsyncAction, ReduxAction } from 'app/common/models/actions';
import { BuildTypeDev } from 'app/common/models/build_type';
import { RootReducers, RootState } from 'app/common/models/global_state';
import { useDispatch } from 'react-redux';
import { AnyAction, compose, Dispatch, Middleware, MiddlewareAPI, Reducer, Store } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

interface AppState {
  [key: string]: object;
}
type RootStateKeys = keyof RootState;
type AppliedMiddleware = (next: (action: unknown) => unknown) => (action: unknown) => unknown;
interface RegisterReducerAction<T extends string = typeof REGISTER_REDUCER> {
  name: RootStateKeys;
  type: T;
  [extraProps: string]: unknown;
}
const REGISTER_REDUCER = 'app/0';
const createDynamicMiddlewares = () => {
  const apiMiddlewares: { [key: string]: Middleware } = {};
  const appliedApiMiddlewares: { [key: string]: AppliedMiddleware } = {};
  let store: MiddlewareAPI<Dispatch<AnyAction>, any>;

  const enhancer: Middleware = (_store) => {
    store = _store;
    return (next: (action: unknown) => unknown) => (action: any) => {
      return (
        compose(...Object.values(appliedApiMiddlewares)) as (
          _next: (_action: unknown) => unknown,
        ) => (_action: unknown) => unknown
      )(next)(action);
    };
  };

  const addMiddleware = (key: string, middleware: Middleware) => {
    appliedApiMiddlewares[key] = middleware(store);
    apiMiddlewares[key] = middleware;
  };

  return {
    enhancer,
    addMiddleware,
  };
};

const dynamicMiddlewaresInstance = createDynamicMiddlewares();

export interface IAppStore {
  registerReducer: <K extends keyof RootState>(
    key: K,
    reducer: RootReducers[keyof RootReducers],
  ) => void;
  getReduxStore: () => Store<AppState, AnyAction>;
  getReduxStoreState: () => RootState;
  dispatchAction: <T extends string>(action: ReduxAction<T> | AsyncAction) => any;
}

export const AppStore = (config: Config, preloadedState?: Partial<RootState>): IAppStore => {
  const reducers: {
    [key: string]: Reducer<{}> | Slice<AppState, SliceCaseReducers<AppState>, RootStateKeys>;
  } = {};

  /**
   * The rootReducer calls each registered reducer with its state key after iterating over all of them,
   * much to combineReducers but with the flexibility to take a registration pattern.
   */
  const _rootReducer = (state: AppState, action: RegisterReducerAction) => {
    if (!state) {
      // eslint-disable-next-line no-param-reassign
      state = {};
    }

    let nextState: AppState = {};
    let hasChanged = false;

    if (action.type === REGISTER_REDUCER && typeof reducers[action.name] === 'function') {
      let reducer = reducers[action.name];
      if (typeof reducer !== 'function') reducer = reducer.reducer;
      const newState = reducer(state[action.name], {
        type: '@@INIT',
      });
      if (newState === undefined) {
        throw new Error(`Reducer for key ${action.name} returned undefined!`);
      }
      nextState = { ...state, [action.name]: newState };
      hasChanged = true;
    } else {
      for (const key in reducers) {
        if (reducers[key]) {
          let reducer = reducers[key];
          if (typeof reducer !== 'function') reducer = reducer.reducer;
          const newState = reducer(state[key], action);
          if (newState === undefined) {
            throw new Error(`Reducer for key ${key} returned undefined!`);
          }
          nextState[key] = newState;
          hasChanged = hasChanged || newState !== state[key];
        }
      }
    }
    return hasChanged ? nextState : state;
  };
  const store = configureStore({
    reducer: (_rootReducer as Reducer<AppState, RegisterReducerAction>) || {},
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        // .concat(this._sagaMiddleware)
        .concat(dynamicMiddlewaresInstance.enhancer),
    devTools: config.buildType === BuildTypeDev,
    preloadedState,
  });
  /**
   * Dispatch a Redux action.
   */
  const dispatchAction = <T extends string>(action: ReduxAction<T> | AsyncAction) =>
    store.dispatch(action as any);

  /**
   * Registers a new reducer on the state.
   *
   * @param key The state key to store the reducer's state under.
   * @param reducer The reducer itself.
   */
  const registerReducer = <K extends RootStateKeys>(
    key: K,
    // it can't be RootReducers[K] because i keep getting "Expression produces a union type that is too complex to represent" error.
    reducer: RootReducers[keyof RootReducers],
    // sagas to register
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ) => {
    if (reducers[key]) return;
    reducers[key] = reducer as unknown as Reducer<{}, AnyAction>;
    // Initialize newly registered reducers by firing an action
    dispatchAction({
      type: REGISTER_REDUCER,
      name: key,
    } as RegisterReducerAction);
  };

  const getReduxStore = () => store as {} as Store<AppState>;

  /**
   * Returns the current Redux state.
   */
  const getReduxStoreState = () => store.getState() as {} as RootState;
  return {
    registerReducer,
    getReduxStore,
    getReduxStoreState,
    dispatchAction,
  };
};

export const useAppDispatch: () => ThunkDispatch<RootState, undefined, AnyAction> = useDispatch;
