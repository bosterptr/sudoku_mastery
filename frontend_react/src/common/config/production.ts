import { Config } from 'app/common/config/config';
import { Locale, LocaleData } from 'app/core/i18n';
import { PRODUCTION_API_BASE_URL, PRODUCTION_BASE_URL } from './constants';
import { BuildTypeProduction } from '../models/build_type';

/**
 * Configuration overrides for the production environment.
 */
export class ProductionConfig extends Config {
  public buildType = BuildTypeProduction;

  public baseURL = PRODUCTION_BASE_URL;

  public apiBaseURL = PRODUCTION_API_BASE_URL;

  public authSettings = {
    /** Name of the cookie to store user authentication in. */
    cookieName: 'auth',
  };

  public locales: Locale[] = [
    {
      name: 'English',
      languageCode: 'en',
      locale: 'en-US',
      default: true,
    },
    {
      name: 'Polski',
      languageCode: 'pl',
      locale: 'pl-PL',
      loader: () =>
        import(
          /* webpackChunkName: 'core.locales.pl-PL' */ 'app/core/i18n/locales/pl-PL'
        ) as unknown as Promise<LocaleData>,
    },
  ];
}
