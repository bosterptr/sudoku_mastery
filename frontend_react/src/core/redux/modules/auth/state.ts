/* eslint-disable no-underscore-dangle */
/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from 'app/common/models/global_state';
import { app } from 'app/core/App';
import { APIState, IAuthAPIStateKeys, IAuthState } from './types';

export const AUTH_NAMESPACE_KEY = 'auth';
const initialState: IAuthState = {};
export type IAuthPayloadAction = PayloadAction<{
  key: IAuthAPIStateKeys;
  state: APIState;
}>;
export const AuthSlice = createSlice({
  name: AUTH_NAMESPACE_KEY,
  initialState,
  reducers: {
    setAPIState(draft, action: PayloadAction<{ key: IAuthAPIStateKeys; state: APIState }>) {
      draft[action.payload.key] = action.payload.state;
    },
  },
});

app.store.registerReducer(AUTH_NAMESPACE_KEY, AuthSlice);

export const selectAuthAPIState = (state: RootState, key: IAuthAPIStateKeys) =>
  state.auth[key] || {
    done: false,
    hasErrored: false,
    isLoading: false,
  };
