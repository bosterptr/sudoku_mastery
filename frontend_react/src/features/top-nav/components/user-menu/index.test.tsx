/* eslint-disable react/jsx-props-no-spreading */
import { ProductionConfig } from 'app/common/config/production';
import { AppStore } from 'app/core/app_store';
import SessionActions from 'app/core/redux/modules/session/actions';
import { SESSION_NAMESPACE_KEY, SessionSlice } from 'app/core/redux/modules/session/state';
import { render } from 'app/tests/utils';
import { UserMenu } from '.';

describe('UserMenu', () => {
  describe('when the user is authenticated', () => {
    it('renders correctly and matches snapshot', () => {
      const store = AppStore(new ProductionConfig());
      store.registerReducer(SESSION_NAMESPACE_KEY, SessionSlice);
      const { asFragment } = render(<UserMenu />, { store });
      expect(asFragment()).toMatchSnapshot();
    });
    it('displays sign in link when not authenticated', () => {
      const store = AppStore(new ProductionConfig());
      store.registerReducer(SESSION_NAMESPACE_KEY, SessionSlice);
      const { queryByText } = render(<UserMenu />, { store });
      expect(queryByText(/sign in/i)).toBeInTheDocument();
      expect(queryByText('d')).not.toBeInTheDocument(); // First letter of displayName
    });
    test('displays user avatar when authenticated', () => {
      const store = AppStore(new ProductionConfig());
      const user = {
        id: '1',
        email: 'email@email.email',
        displayName: 'displayName',
      };
      store.registerReducer(SESSION_NAMESPACE_KEY, SessionSlice);
      store.dispatchAction(SessionActions.setSessionUser(user));
      const { queryByText } = render(<UserMenu />, { store });
      expect(queryByText(/sign in/i)).not.toBeInTheDocument();
      expect(queryByText('d')).toBeInTheDocument(); // First letter of displayName
    });
  });
});
