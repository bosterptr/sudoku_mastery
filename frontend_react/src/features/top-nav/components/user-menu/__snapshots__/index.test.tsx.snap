// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu when the user is authenticated renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  .c1 {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: flex-end;
  width: 100%;
}

.c2 {
  background: #1648ff;
  border-radius: 1rem;
  color: #fff;
  cursor: pointer;
  font-weight: 600;
  margin: auto 10px;
  padding: 0.9rem 1rem;
  transition: all 0.3s ease;
}

.c2:hover {
  border-radius: 2rem;
  color: rgb(229, 229, 229);
}

.c0 {
  z-index: 1000;
  height: 50px;
  flex-shrink: 0;
}

<nav
    class="c0"
  >
    <div
      class="c1"
    >
      <a
        class="c2"
        href="/auth/signin"
      >
        Sign in
      </a>
    </div>
  </nav>
</DocumentFragment>
`;
